### DISCUSSION TITLE
Finetune Kimi K2.5 VL in NeMo Automodel

### DISCUSSION THREAD

**Discussion by HuiyingLi** [NVIDIA-NeMo/Automodel#1161] [Show and tell]
[`moonshotai/Kimi-K2.5`](https://huggingface.co/moonshotai/Kimi-K2.5) is Moonshot AI’s flagship open-source “native multimodal agentic” model: a large Mixture-of-Experts Transformer (1T total parameters with ~32B activated per token) designed to handle text plus vision inputs while also acting as a tool-using agent with both fast “instant” and deeper “thinking” styles of response.

### Parallel Setup
We provide a fine-tuning [recipe](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/vlm_finetune/kimi/kimi25vl_medpix.yaml) for Kimi K2.5 that scales training using Expert Parallelism and Pipeline Parallelism. The configuration runs with EP=32 and PP=8 across 32 nodes (8× H100 GPUs per node).
```yaml
distributed:
  _target_: nemo_automodel.components.distributed.fsdp2.FSDP2Manager
  tp_size: 1
  cp_size: 1
  pp_size: 8
  dp_replicate_size: 1
  ep_size: 32
  sequence_parallel: false

autopipeline:
  _target_: nemo_automodel.components.distributed.pipelining.AutoPipeline
  pp_schedule: interleaved1f1b
  pp_microbatch_size: 1
  round_virtual_stages_to_pp_multiple: down
  scale_grads_in_schedule: false
  layers_per_stage: 2
  patch_inner_model: false      
  patch_causal_lm_model: false 
```
### Data
We use the [MedPix-VQA](https://huggingface.co/datasets/mmoukouba/MedPix-VQA) dataset as an example. MedPix-VQA is a medical visual question answering dataset for training and evaluating VQA models in the clinical imaging domain, pairing radiological images from MedPix (a well-known medical image collection) with question–answer examples that target medical image interpretation.

Below is the loss curve obtained when fine-tuning on MedPix-VQA with this recipe:
<img width="560" height="309" alt="image" src="https://github.com/user-attachments/assets/ceeb5811-2ab1-4569-8fa0-008ed6f5e8d5" />

**Comment by Bambuuai** [comment]
Thanks @HuiyingLi 

What is the minimum resource requirements for this?

**Reply by HuiyingLi** [reply]
Hi! Thanks for interest! This recipe runs with 32 nodes x8 H100.

**Reply by Bambuuai** [reply]
Thanks for your reply!
Is it possible to run with H200x8? while training, which precision is used for weights? bf16 or int4 or fp8?
Is there a recipe for H200 x 8?

**Reply by HuiyingLi** [reply]
Hi @Bambuuai ,
We didn't try H200 x 8, feel free to experiment with the ep/pp size.
The weights are dequant to bf16 for training.

**Reply by Bambuuai** [reply]
I tried and failed because of that. because of dequant to bf16
I am going to implement lora fine-tuning without dequantization, what do you think? is it possible? Of course, i am not merging adapter.

**Reply by HuiyingLi** [reply]
Hi @Bambuuai  could you please share your err messages with the dequantization? Could you try our r0.3.0 branch? Thank you!

**Reply by Bambuuai** [reply]
It's OOM error, I used H200x8
If this recipe use bf16 for lora, H200x8 is not enough for that.

**Reply by Bambuuai** [reply]
Also, kimi tokenizer in hf is not proper for reasoning model training, which chat_template is used here?

**Reply by HuiyingLi** [reply]
This recipe uses the hf chat template. Could you please describe what's your use case and dataset you are looking at? Thank you!

**Reply by Bambuuai** [reply]
<img width="893" height="941" alt="image" src="https://github.com/user-attachments/assets/cd230b0e-765f-43ad-8445-f5a551c050c3" />

In hf chat template, last assistant message is included in `hist_msgs` . and `hist_msgs` doesn't include `reasoning_content`.
I think it is not suitable for post-training, ;)

**Reply by HuiyingLi** [reply]
@Bambuuai  Thank you for your suggestion! We can expose the chat template for user to pass in their own chat template.

**Reply by Bambuuai** [reply]
Hi @HuiyingLi 

I made PR about this part. If you have a moment to review or if there are specific adjustments needed to get this merged, please let me know. Thank you

**Reply by Bambuuai** [reply]
https://github.com/NVIDIA-NeMo/Automodel/pull/1525
This is pr link

**Reply by HuiyingLi** [reply]
@Bambuuai  Thank you so much for your contribution!! 🚀

**Reply by Bambuuai** [reply]
Thanks @HuiyingLi, could you review my PR please?

**Reply by Bambuuai** [reply]
Hi @HuiyingLi 
Thank you for your approval. But the PR is still not merged, could you let me know what I should adjust more please?

**Reply by HuiyingLi** [reply]
Looks like it's still in queue in ci. If it still didn't get pass today we will check if anything is up with ci. Thanks!

**Reply by Bambuuai** [reply]
Thanks. it merged
